File: src/main/java/avalon/group/Execute.java
package avalon.group;

import avalon.extend.Executive;
import avalon.model.ExtendLanguage;
import avalon.tool.pool.ExtendLanguagePool;
import avalon.util.GroupMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

import static avalon.tool.Responder.AT;

/**
 * Created by Eldath Ray on 2017/6/11 0011.
 *
 * @author Eldath Ray
 */
public class Execute implements GroupMessageResponder {
    private static final Logger logger = LoggerFactory.getLogger(Execute.class);
    private static final Set<Pattern> dangerCommand = new HashSet<>();
    private static final Execute instance = new Execute();

    static {
        dangerCommand.add(Pattern.compile("shutdown"));
    }

    @Override
    public void doPost(GroupMessage message) {
        String content = message.getContent();
        String[] lines = content.split("\n");
        String[] split = lines[0].split(" ");
        if (split.length < 3) {
            message.response(AT(message) + " 指令不合法 ⊙﹏⊙!");
            return;
        }
        for (Pattern thisDangerCommand : dangerCommand) {
            for (String thisLine : lines)
                if (thisDangerCommand.matcher(thisLine.trim()
                        .toLowerCase()
                        .replaceAll("[\\pP\\p{Punct}]", "")).matches()) {
                    message.response(AT(message) + " 代码中包含危险字，不允许执行！若有疑问请艾特Eldath~");
                    return;
                }
        }
        StringBuilder code = new StringBuilder();
        for (int i = 1; i < lines.length; i++)
            code.append(lines[i]).append("\n");

        String languageString = split[2].trim().toLowerCase();
        ExtendLanguage lang = ExtendLanguagePool.get(languageString);
        if (lang != null) {
            Map<String, Object> result;
            try {
                result = Executive.execute(lang, code.toString());
            } catch (Exception e) {
                logger.warn("exception thrown while execute code: " + e.toString());
                message.response("Oh no! 代码执行失败：未知错误");
                return;
            }
            String errorOrOut = (String) result.get("out/error");
            if ((boolean) result.get("error"))
                message.response(AT(message) + " Oops! 您的代码执行错误! 错误如下：\n" + errorOrOut);
            else
                message.response(AT(message) + " Yap! 提交的代码已经执行完成。执行输出：\n" + errorOrOut);
        } else {
            message.response(AT(message) + " 指定的语言目前暂不支持~ 抱歉`(*>﹏<*)′");
        }
    }

    public static Execute getInstance() {
        return instance;
    }

    @Override
    public String getHelpMessage() {
        return "avalon execute <语言，目前仅支持Python>{换行}<程序，支持多行>：执行指定语言的程序并返回输出";
    }

    @Override
    public Pattern getKeyWordRegex() {
        return Pattern.compile("avalon execute \\w+");
    }

    @Override
    public GroupMessageResponder instance() {
        return null;
    }
}

File: src/main/kotlin/avalon/group/Execute.kt
package avalon.group

import avalon.api.Flag
import avalon.api.Flag.at
import avalon.tool.ExecutiveStatus
import avalon.tool.Executives
import avalon.tool.pool.Constants.Basic.LANG
import avalon.tool.pool.Constants.Basic.MAX_ECHO_LENGTH
import avalon.util.GroupConfig
import avalon.util.GroupMessage
import java.util.regex.Pattern

object Execute : GroupMessageResponder() {
	private val executive = Executives.EXECUTIVE

	override fun doPost(message: GroupMessage, groupConfig: GroupConfig) {
		val contentM = message.content
		val single = !contentM.contains("\n")

		val codes: List<String>
		val lang: String
		if (single) {
			val split = contentM.split(" ") // avalon execute python print(2+3)
			val size = split.size
			if (size < 4) {
				message.response("${at(message)} ${LANG.getString("group.execute.incorrect")}")
				return
			}
			codes = split.subList(3, size)
			lang = split[2]
		} else {
			val split = contentM.split("\n")
			val first = split[0]
			codes = split.subList(1, split.size)
			lang = first.replace("avalon execute ", "")
				.replace("avalon ex ", "").trim()
		}

		if (!executive.allLanguages().contains(lang)) {
			message.response("${at(message)} ${LANG.getString("group.execute.unsupported_language")}")
			return
		}
		if (codes.all { it.isEmpty() }) {
			message.response("${at(message)} ${LANG.getString("group.execute.empty_code")}")
			return
		}
		val result = executive.execute(lang, codes)
		val exitcode = result.exitcode
		val stderr = handleOutput(result.stderr)
		val content =
			when (result.status) {
				ExecutiveStatus.ERROR -> LANG.getString("group.execute.error")
					.format("exitcode: $exitcode} stderr: $stderr error: ${handleOutput(result.error)}")
				ExecutiveStatus.STDERR -> LANG.getString("group.execute.stderr")
					.format("exitcode: $exitcode stderr: $stderr")
				ExecutiveStatus.OK -> LANG.getString("group.execute.ok")
					.format("exitcode: $exitcode stdout: ${handleOutput(result.stdout)}")
			}
		message.response("${Flag.at(message)} $content")
	}

	private fun handleOutput(string: String): String {
		val length = string.length
		if (length > MAX_ECHO_LENGTH)
			return string.substring(0, MAX_ECHO_LENGTH) + "...${String.format(LANG.getString("group.execute.truncated"), length)}"
		return string.replace("\n", " ")
	}

	override fun responderInfo(): ResponderInfo =
		ResponderInfo(
			Pair("(ex|execute) ${LANG.getString("group.execute.help.first")}",
				LANG.getString("group.execute.help.second").format("avalon execute info")),
			Pattern.compile("(ex|execute) [^info]*")
		)

	override fun instance(): GroupMessageResponder = this
}